lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -254,43 +254,49 @@ public void run() {
 	 * transport events/notifications
 	 ***************************************/
 
-	void onMessage(ProtocolMessage message) {
-		if(protocolListener != null)
-			protocolListener.onRawMessage(message);
-		switch(message.action) {
-		case HEARTBEAT:
-			onHeartbeat(message);
-			break;
-		case ERROR:
-			ErrorInfo reason = message.error;
-			if(reason == null)
-				Log.e(TAG, "onMessage(): ERROR message received (no error detail)");
-			else
-				Log.e(TAG, "onMessage(): ERROR message received; message = " + reason.message + "; code = " + reason.code);
+	void onMessage(ProtocolMessage message) throws AblyException {
```

Comment:
@paddybyers @mattheworiordan 

TL,DR, 

I am strongly advocating against handling errors that caused by non-client-caused reasons. (e.g., provided message isn't compatible with our protocol, breaking our secred wow between client and server)


---

I can handle the thrown exception within the `ConnectionManager#onMessage` scope, but before that I would like to mention couple of things.



> An error in processing a correctly deserialised message, I think, should be handled here.

Like you said `that catch was intended to be handling any deserialisation failure of the message`, and we don't have a deserialisation issue here. What we are trying to handle here is a message that is not compatible with our protocol. If we try to handle that (and whole possible non-client-caused errors that might occur) using either,

  - validation, which introduces following side-effects
    - boilerplate code. Proof of concept, https://github.com/gokhanbarisaker/ably-java/commit/ba1564ca24e3b85af716316294c09de8f5210908
    - performance overhead, (this is a blind guess made without profiling code, actually a little arrogant of me to add it here without it :))
    - a little bit garbage creation.

  - try/catch blocks, which introduces following side-effects
    - lots of boilerplate code,
    - performance overhead (this is a blind guess made without profiling code, actually a little arrogant of me to add it here without it :))
    - although current handling for unchecked exceptions on upper scope just logs the data, trying to handle that specific exception can be easily confused with others.


We probably end up violating KISS principle, and arguably introduce over engineering. Also above side-effects multiply considering out client will process high throughput during realtime messaging.



If you are still OK with the things above, I am comfortable with handling the exception within current scope :)